### PR TITLE
Automate include grouping order in .clang-format

### PR DESCRIPTION
**Pull request by harrism**
## Description
This uses the `IncludeCategories` settings in .clang-format to attempt to enforce our documented `#include` order in libcudf. See https://docs.rapids.ai/api/libcudf/stable/developer_guide

I realize that there was a [previous attempt at this](https://github.com/rapidsai/cudf/pull/12760) by @bdice that met with some resistance. Reading it, I wouldn't say it was vetoed; rather, reviewers requested something much simpler. I have a few reasons to attempt this again.

1. To make a separate task much easier. We are undertaking a refactoring of RMM that will replace `rmm::mr::device_memory_resource*` with `rmm::device_async_resource-ref` everywhere in RAPIDS (not just cuDF). This requires adding an include to MANY files. Getting the location of the include right everywhere is very difficult without automatic grouping of headers. I started out writing a bash script to do this before realizing clang-format has the necessary feature. And I realized that my script would never properly handle [files like this](https://github.com/rapidsai/raft/blob/branch-24.04/cpp/bench/ann/src/raft/raft_cagra_wrapper.h).
2. To increase velocity. Everywhere in RAPIDS that we have automated code standard/style/formatting/other, the benefits to velocity have outweighed the costs. To paraphrase @bdice, $auto \nearrow \rightarrow \mu \searrow \rightarrow v \nearrow$
3. The previous PR #12760 had nearly 50 categories of headers. There was no way this could be applied universally across RAPIDS repos. My proposal has 10 categories. I tried to reduce it further but realized that it wouldn't be much less configuration to maintain, so I stopped at 10.

Note that one of the ways that having few categories can work while still maintaining clear groups is that this PR updates many files to use quotes ("") instead of angle brackets (<>) for local cuDF headers that do not live in `cudf/cpp/include`. With our "near to far" include ordering policy, these are arguably the nearest files, and using quotes allows us to have our first category simply check for quotes. These files will be grouped and sorted without blank lines, but in practice this does not lose clarity because typically headers from more than two directories are not included from the same file. The downside of this change is I don't yet know how to automatically enforce it. I hope that when developers accidentally use <> for internal includes that don't start with (e.g.) "cudf", they will be grouped one of the lowest priority categories, and perhaps this will induce them to switch to "" to get the headers listed at the top. The rule is simple: if it's in libcudf but not in `cpp/include/cudf`, then use quotes. For **everything** else, use angle brackets.

Other than headers from RAPIDS repos, I group all other headers that have a file extension in a single group, and all files that have no file extension in another group. Since the latter also matches includes some files from libcu++, I have an explicit category for `<cuda/` includes to keep them separate from STL includes. A frequent effect of the single "." group is that cub and thrust headers get grouped without a blank line between them. I don't think this is a problem.

Below I'm listing the (fairly simple, in my opinion) .clang-format settings for this PR. Note that categories 2-5 will require tweaking for different RAPIDS repos. 

Some may ask why I ordered `cudf_test` headers before `cudf` headers. I tried both orders, and putting `cudf_test` first generated significantly fewer changes in the PR, meaning that it's already the more common ordering (I suppose `cudf_test` is closer to the files that include it, since they are libcudf tests).

I've opened a similar PR for RMM with only 5 groups. https://github.com/rapidsai/rmm/pull/1463

CC @davidwendt @vyasr @wence- @GregoryKimball for feedback

@isVoid contributed to this PR via pair programming.

```
IncludeBlocks: Regroup
IncludeCategories:
  - Regex:           '^"' # quoted includes
    Priority:        1
  - Regex:           '^<(benchmarks|tests)/' # benchmark includes
    Priority:        2
  - Regex:           '^<cudf_test/' # cuDF includes
    Priority:        3
  - Regex:           '^<(nvtext|cudf_kafka)' # other libcudf includes
    Priority:        4
  - Regex:           '^<cudf/' # cuDF includes
    Priority:        5
  - Regex:           '^<(cugraph|cuml|cuspatial|raft|kvikio)' # Other RAPIDS includes
    Priority:        6
  - Regex:           '^<rmm/' # RMM includes
    Priority:        7
  - Regex:           '^<(thrust|cub|cuda)/' # CCCL includes
    Priority:        8
  - Regex:           '^<(cuda.h|cuda_runtime|device_types|math_constants|cooperative_groups)' # CUDART includes
    Priority:        8
  - Regex:           '^<.*\..*' # other system includes (e.g. with a '.')
    Priority:        9
  - Regex:           '^<[^.]+' # STL includes (no '.')
    Priority:        10
```

## Checklist
- [X] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [X] New or existing tests cover these changes.
- [X] The documentation is up to date with these changes.

### CODE REVIEW

File: cpp/benchmarks/join/generate_input_tables.cuh [hunk 1]
```diff
@@ -22,13 +22,12 @@
 
 #include <rmm/exec_policy.hpp>
 
+#include <curand.h>
+#include <curand_kernel.h>
 #include <thrust/distance.h>
 #include <thrust/iterator/counting_iterator.h>
 #include <thrust/sequence.h>
```
Comment (isVoid) [inline cpp/benchmarks/join/generate_input_tables.cuh:29]: Should CCCL libraries be its own category too? Since we use them pervasively, placing them in a separate group help identifying if there are unused headers in the future.

**Comment by harrism** [comment]
Note that these changes did turn up two places where needed headers were not included.

File: cpp/benchmarks/join/generate_input_tables.cuh (hunk 1 above)
Comment (bdice) [reply cpp/benchmarks/join/generate_input_tables.cuh:29]: My vote would be:
- Thrust/CUB in its own category
- Other "dot" headers from the system
- libcudacxx `<cuda/...>`
- STL (no dot)

Comment (bdice) [reply cpp/benchmarks/join/generate_input_tables.cuh:29]: I would also accept:
- CCCL (Thrust/CUB/libcudacxx) in its own category
- Other "dot" headers from the system
- STL (no dot)

Comment (davidwendt) [reply cpp/benchmarks/join/generate_input_tables.cuh:29]: I also vote for `cuda/, thrust/, cub/` in a single category

Comment (harrism) [reply cpp/benchmarks/join/generate_input_tables.cuh]: Done. I like this too. I also put cuco and CUDA headers (cuda_runtime.h etc.) in the same group. This adds one more regex (11), but same number of groups (10).

**Review by bdice — COMMENTED** [review]
Flushing comments for now. I'll look further at this later.

**Comment by davidwendt** [comment]
Seems something here makes the jitify code-paths fail. I know jitify has issues with certain include files.
I would recommend addressing the compile errors sooner than later.

**Comment by harrism** [comment]
It built fine locally. Will look into it.

File: cpp/src/transform/row_conversion.cu [hunk 2]
```diff
@@ -32,15 +31,19 @@
...
+#include <cooperative_groups.h>
 #include <thrust/binary_search.h>
 #include <thrust/iterator/counting_iterator.h>
 #include <thrust/iterator/discard_iterator.h>
 #include <thrust/iterator/transform_iterator.h>
 #include <thrust/scan.h>
+
 #include <type_traits>
 
 #if defined(__CUDA_ARCH__) && __CUDA_ARCH__ >= 700
```
Comment (bdice) [inline cpp/src/transform/row_conversion.cu:49]: We require Volta+ now, so we can probably remove this part of the macro?
```suggestion
#if defined(__CUDA_ARCH__)
```

File: cpp/tests/interop/from_arrow_test.cpp [hunk 3]
```diff
@@ -14,6 +14,14 @@
  * limitations under the License.
  */
 
+#include <tests/interop/arrow_utils.hpp>
```
Comment (bdice) [inline cpp/tests/interop/from_arrow_test.cpp:17]: What is our heuristic for whether files in `tests/` should use local includes? It seems like we might have inconsistency here but I didn't check that closely.

File: cpp/tests/io/comp/decomp_test.cpp [hunk 4]
```diff
@@ -14,15 +14,16 @@
  * limitations under the License.
  */
 
-#include <io/comp/gpuinflate.hpp>
-#include <io/utilities/hostdevice_vector.hpp>
-#include <src/io/comp/nvcomp_adapter.hpp>
-
-#include <cudf/utilities/default_stream.hpp>
+// TODO : don't include src from tests!
```
Comment (bdice) [inline cpp/tests/io/comp/decomp_test.cpp:17]: Adding a comment to track this TODO. It might be easiest to solve this in a separate PR, but I'm not 100% sure what the outcome should be. There seem to be quite a few test files, especially I/O tests, that rely on files from `src`.

**Review by bdice — COMMENTED** [review]
A few comments -- I reviewed the entire PR, and my comments are fairly minor, so I am eager to move this forward largely as-is.

File: cpp/benchmarks/io/orc/orc_writer_chunks.cpp
[the remark this answers is not in the document]
Comment (harrism) [reply cpp/benchmarks/io/orc/orc_writer_chunks.cpp:1]: Is there a way to find files that only have copyright changes? Is there really a problem if we update the copyright on a file? After all, changing the copyright date is a change to the file, warranting a change of copyright date. :)

File: cpp/src/io/parquet/reader_impl_helpers.cpp [hunk 5]
```diff
@@ -16,7 +16,7 @@
 
 #include "reader_impl_helpers.hpp"
 
-#include <io/utilities/row_selection.hpp>
+#include "io/utilities/row_selection.hpp"
```
[the remark this answers is not in the document]
Comment (harrism) [reply cpp/src/io/parquet/reader_impl_helpers.cpp:19]: I can't figure it out either.  I can't see any other regex that this would match.  But I can't get it to group it any other way.

[the remark this answers is not in the document]
Comment (harrism) [reply cpp/src/io/parquet/reader_impl_helpers.cpp:19]: If I add any more quoted headers they get grouped correctly. If I remove this header and add a different header without directories, it gets separated like this one. I suspect a clang-format bug.

**Comment by wence-** [comment]
I am very much in favour of this. I like the groupings. One minor question around the colour of the bikeshed:

Do we want to (probably not in this change due to the churn it would cause) have separate groups for the public (`<cudf/public_header.hpp>`) and detail (`<cudf/detail/private_header.hpp>`) include separation?

File: cpp/src/transform/row_conversion.cu (hunk 2 above)
Comment (wence-) [reply cpp/src/transform/row_conversion.cu:49]: Let us please not do this here. This change is really substantively only a change of include ordering, so if we wanted to make future "git-annotate" usage easier we could add the single commit to the git blame ignore file so that annotations skip over things. This change, which is substantive, is easily lost in such a scenario.

Comment (bdice) [reply cpp/src/transform/row_conversion.cu:49]: Totally agree - just didn’t want to ignore this. A follow-up PR or opening an issue is appropriate.

Comment (harrism) [reply cpp/src/transform/row_conversion.cu:49]: Please open an issue.

File: cpp/tests/interop/from_arrow_test.cpp (hunk 3 above)
Comment (harrism) [reply cpp/tests/interop/from_arrow_test.cpp:17]: Please file an issue.

**Comment by harrism** [comment]
The commit history on this PR has gotten out of hand somehow. I think it's best to start a new PR with the latest agreed-upon .clang-format changes and then rerun the pre-commit hooks and apply any other applicable reviewer suggestions from this PR. This way I will minimize changes and have a better change of figuring out any CI failures.

**Comment by wence-** [comment]
Closing in favour of #15063
